### PR TITLE
(PIX passes) Take constant color from CB

### PR DESCRIPTION
**Pull request by jeffnn**
(This change includes a patch from Xiang that allows me to re-create resource meta data after adding a reference to a CB. Thanks, Xiang!)

The goal here is to extend the override-output-with-constant-color so that we can optionally derive the constant color from a CB rather than just from a literal constant. PIX uses this capability to draw its "pixel cost" visualization.

Sorry the diffs look a bit bad at the top of DxilOutputColorBecomesConstant.cpp - I pulled out the loop over the calls to storeOutput and pass it a lambda. Not much actual code change there.

### CODE REVIEW

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp [hunk 1]
```diff
@@ -150,18 +133,128 @@ bool DxilOutputColorBecomesConstant::runOnModule(Module &M)
...
+    hasIntOutputs = true;
+  });
+
+    if (!hasFloatOutputs && !hasIntOutputs)
+  {
+    return false;
+  }
+
+  // Otherwise, we assume the shader outputs only one or the other (because the 0th RTV can't have a mixed type)
+  assert(hasFloatOutputs || hasIntOutputs);
```
Comment (python3kgae) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: !hasFloatOutputs || !hasIntOutputs ?
Also use DXASSERT( condition, msg) for assert.

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp [hunk 2]
```diff
@@ -150,18 +133,128 @@ bool DxilOutputColorBecomesConstant::runOnModule(Module &M)
...
+      Constant *CreateHandleOpcodeArg = HlslOP->GetU32Const((unsigned)DXIL::OpCode::CreateHandle);
+      Constant *CBVArg = HlslOP->GetI8Const(static_cast<std::underlying_type<DxilResourceBase::Class>::type>(DXIL::ResourceClass::CBuffer)); 
+      Constant *MetaDataArg = HlslOP->GetU32Const(0); // position of the metadata record in the corresponding metadata list
+      Constant *IndexArg = HlslOP->GetU32Const(0); // 
+      Constant *FalseArg = HlslOP->GetI1Const(0); // non-uniform resource index: false
+      CallInst *callCreateHandle = Builder.CreateCall(createHandle, { CreateHandleOpcodeArg, CBVArg, MetaDataArg, IndexArg, FalseArg }, ConstantBufferName);
+
+      pCBuf->SetHandle(callCreateHandle);
+
+      ID = DM.AddCBuffer(std::move(pCBuf));
```
Comment (python3kgae) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: Move ID = DM.AddCBuffer(std::mov(pCBuf)); up so 
MetaDataArg could use the ID.
Constant *MetaDataArg = HlslOP->GetU32Const(ID); //

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp [hunk 3]
```diff
@@ -150,18 +133,128 @@ bool DxilOutputColorBecomesConstant::runOnModule(Module &M)
...
+      auto ConstantBufferName = "PIX_Constant_Color_CB_Handle";
+      Function *createHandle = HlslOP->GetOpFunc(DXIL::OpCode::CreateHandle, Type::getVoidTy(Ctx));
+      Constant *CreateHandleOpcodeArg = HlslOP->GetU32Const((unsigned)DXIL::OpCode::CreateHandle);
+      Constant *CBVArg = HlslOP->GetI8Const(static_cast<std::underlying_type<DxilResourceBase::Class>::type>(DXIL::ResourceClass::CBuffer)); 
+      Constant *MetaDataArg = HlslOP->GetU32Const(0); // position of the metadata record in the corresponding metadata list
+      Constant *IndexArg = HlslOP->GetU32Const(0); // 
+      Constant *FalseArg = HlslOP->GetI1Const(0); // non-uniform resource index: false
+      CallInst *callCreateHandle = Builder.CreateCall(createHandle, { CreateHandleOpcodeArg, CBVArg, MetaDataArg, IndexArg, FalseArg }, ConstantBufferName);
+
+      pCBuf->SetHandle(callCreateHandle);
```
Comment (python3kgae) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: Don't need to set handle.

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp [hunk 4]
```diff
@@ -150,18 +133,128 @@ bool DxilOutputColorBecomesConstant::runOnModule(Module &M)
...
+      ConstantInt * OutputColumnConstant = cast<ConstantInt>(OutputColumnOperand);
+      APInt OutputColumn = OutputColumnConstant->getValue();
+      CallInstruction->setOperand(hlsl::DXIL::OperandIndex::kStoreOutputValOpIdx, ReplacementColors[*OutputColumn.getRawData()]);
+    });
   }
 
-  Function * IntOutputFunction = HlslOP->GetOpFunc(DXIL::OpCode::StoreOutput, Type::getInt32Ty(Ctx));
-  if (IntOutputFunction->getNumUses() != 0) {
-    Modified = convertTarget0ToConstantValue(IntOutputFunction, OutputSignature, HlslOP, color);
+  if (!IntOutputFunction->user_empty()) {
```
Comment (python3kgae) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: if (hasIntOutputs)?

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp [hunk 5]
```diff
@@ -150,18 +133,128 @@ bool DxilOutputColorBecomesConstant::runOnModule(Module &M)
...
+  }
+
   bool Modified = false;
 
   // The StoreOutput function can store either a float or an integer, depending on the intended output
   // render-target resource view.
-  Function * FloatOutputFunction = HlslOP->GetOpFunc(DXIL::OpCode::StoreOutput, Type::getFloatTy(Ctx));
-  if (FloatOutputFunction->getNumUses() != 0) {
-    Modified = convertTarget0ToConstantValue(FloatOutputFunction, OutputSignature, HlslOP, color);
+  if (!FloatOutputFunction->user_empty()) {
```
Comment (python3kgae) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: if (hasFloatOutputs)?

File: tools/clang/test/HLSL/pix/constantcolorFromCBint.hlsl [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+// CHECK: %PIX_Constant_Color_Value2 = extractvalue %dx.types.CBufRet.i32 %PIX_Constant_Color_Value, 2
+// CHECK: %PIX_Constant_Color_Value3 = extractvalue %dx.types.CBufRet.i32 %PIX_Constant_Color_Value, 3
+
+// Check that the store-output has been modifed:
+// CHECK: call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 0, i32 %PIX_Constant_Color_Value0)
+// CHECK: call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 1, i32 %PIX_Constant_Color_Value1)
+// CHECK: call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 2, i32 %PIX_Constant_Color_Value2)
+// CHECK: call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 3, i32 %PIX_Constant_Color_Value3)
+
+[RootSignature("")]
```
Comment (python3kgae) [inline tools/clang/test/HLSL/pix/constantcolorFromCBint.hlsl]: Use a cbuffer to test ID for PIX_Constant_Color_CB not 0.

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp (hunk 1 above)
Comment (jeffnn) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: Ha ha. Yes, you're right. I was thinking in human language: "or" usually means "either but not both" when we say it out loud :-)

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp (hunk 2 above)
Comment (jeffnn) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: Ah, I see. I didn't understand what that meta-data ID meant... Thanks.

File: tools/clang/test/HLSL/pix/constantcolorFromCBint.hlsl (hunk 6 above)
Comment (jeffnn) [inline tools/clang/test/HLSL/pix/constantcolorFromCBint.hlsl]: Great idea.

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp (hunk 4 above)
Comment (jeffnn) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: Thanks.

**Pull request merged**